### PR TITLE
[Mellanox][Smartswitch] Platform API for Smartswtich

### PR DESCRIPTION
**Pull request by gpunathilell**
Should be merged after https://github.com/sonic-net/sonic-buildimage/pull/20943

#### Why I did it
Introduction of `SmartSwitchChassis`, `DpuModule` classes for APIs required for Smartswitch PMON 
- `module.py` - Introduction of `DpuModule` for representing each DPU as a module
- `platform.py` - Initialize `SmartSwitchChassis` based on number of DPUs on platform
- `chassis.py` - Introduction of `SmartSwitchChassis` for representing each the smartswitch as a modular chassis with DPUS present 

##### Work item tracking
- Microsoft ADO **(number only)**:

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

**Review by oleksandrivantsiv — APPROVED** [review]

File: platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py [hunk 1]
```diff
@@ -1198,3 +1198,155 @@ def get_sfp(self, index):
...
+
+    def get_num_modules(self):
+        """
+        Retrieves the number of modules available on this chassis
+        On a SmarSwitch chassis this includes the number of DPUs.
+
+        Returns:
+            An integer, the number of modules available on this chassis
+        """
+        return DeviceDataManager.get_dpu_count()
```
Comment (vvolam) [inline platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py:1254]: This seems to be returning only DPUs?

**Comment by gpunathilell** [comment]
Dependent on for mellanox build to pass https://github.com/sonic-net/sonic-buildimage/pull/20943

File: platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py (hunk 1 above)
Comment (gpunathilell) [reply platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py:1254]: Yes, only the DPUs are represented as modules, the switch itself is not a module

**Review by vvolam — APPROVED** [review]
Looks good to me.

**Comment by vvolam** [comment]
@gpunathilell Please check the build errors.

**Comment by oleksandrivantsiv** [comment]
@kperumalbfn can we merge this?

File: platform/mellanox/mlnx-platform-api/sonic_platform/module.py [hunk 2]
```diff
@@ -247,3 +253,290 @@ def get_sfp(self, index):
...
+        if not self._is_midplane_up():
+            return False
+        command = ['ping', '-c', '1', '-W', '1', self.get_midplane_ip()]
+        try:
+            return subprocess.call(command, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL) == 0
+        except Exception as e:
+            logger.log_error(f"Failed to check midplane reachability for {self.get_name()}! {e}")
+            return False
+
+    def get_midplane_interface(self):
```
Comment (kperumalbfn) [inline platform/mellanox/mlnx-platform-api/sonic_platform/module.py:473]: @gpunathilell Do we have any new sonic-mgmt tests?

Comment (oleksandrivantsiv) [reply platform/mellanox/mlnx-platform-api/sonic_platform/module.py:473]: This is vendor API implementation for Smart Switch PMON: https://github.com/sonic-net/SONiC/blob/master/doc/smart-switch/pmon/smartswitch-pmon.md
The test for the feature will be contributed by Cisco, who is working on the feature implementation.

Comment (kperumalbfn) [reply platform/mellanox/mlnx-platform-api/sonic_platform/module.py:473]: Thanks @oleksandrivantsiv

**Review by kperumalbfn — APPROVED** [review]

**Pull request merged**